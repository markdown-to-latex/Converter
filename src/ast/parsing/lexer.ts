import * as node from '../node';
import {
    CodeNode,
    copyTextPosition,
    createStartEndPos,
    Node,
    NodeType,
    RawNode,
    RawNodeE,
    RawNodeType,
    TextPosition,
    textPositionG,
    textPositionGEq,
} from '../node';
import {
    DiagnoseErrorType,
    DiagnoseList,
    diagnoseListHasSeverity,
    DiagnoseSeverity,
} from '../../diagnose';

export const enum LexemeType {
    // Space,
    // Paragraph,
    // Heading,
    // Text,
    Code,
}

const parserPriorities: LexemeType[] = [
    LexemeType.Code,
    // LexemeType.Heading,
    // LexemeType.Paragraph,
    // LexemeType.Text,
    // LexemeType.Space,
];

interface LexemeTypeToNodeType {
    [LexemeType.Code]: node.CodeNode;
    // [LexemeType.Heading]: node.HeadingNode;
    // [LexemeType.Paragraph]: node.ParagraphNode;
    // [LexemeType.Text]: node.TextNode;
    // [LexemeType.Space]: node.SpaceNode;
}

class FatalError extends Error {
}

export function applyVisitors(nodes: Readonly<Node[]>): [Node[], DiagnoseList] {
    let allDiags: DiagnoseList = [];

    let newNodes = [...nodes];
    try {
        for (const type of parserPriorities) {
            newNodes = newNodes.flatMap(node => {
                if (node.type !== RawNodeType.Raw) {
                    return [node];
                }

                let [nodes, diags] = applyParser(node as RawNode, type);
                allDiags.push(...diags);

                if (diagnoseListHasSeverity(diags, DiagnoseSeverity.Fatal)) {
                    throw new FatalError();
                }

                return nodes;
            });
        }
    } catch (e) {
        if (e instanceof FatalError) {
            return [[], allDiags];
        } else {
            throw e;
        }
    }

    return [newNodes, allDiags];
}

function applyParser(
    node: Readonly<RawNode>,
    type: LexemeType,
): [Node[], DiagnoseList] {
    const nodeE = new RawNodeE(node);

    const diags: DiagnoseList = [];
    const newNodes: Node[] = [];
    let position: TextPosition = copyTextPosition(node.pos.start);

    const parsed = parsers[type](nodeE);
    for (const node of parsed) {
        if (!textPositionGEq(position, node.pos.start)) {
            newNodes.push({
                type: RawNodeType.Raw,
                pos: {
                    start: copyTextPosition(position),
                    end: copyTextPosition(node.pos.start),
                },
                text: nodeE.text.slicePosition(
                    nodeE.n.pos.start,
                    position,
                    node.pos.start,
                ).s,
                parent: nodeE.n.parent,
            } as RawNode);
        }

        newNodes.push(node);
        position = copyTextPosition(node.pos.end);
    }

    if (!textPositionGEq(position, node.pos.end)) {
        newNodes.push({
            type: RawNodeType.Raw,
            pos: {
                start: copyTextPosition(position),
                end: copyTextPosition(node.pos.end),
            },
            text: nodeE.text.slicePosition(
                nodeE.n.pos.start,
                position,
                node.pos.end,
            ).s,
            parent: nodeE.n.parent,
        } as RawNode);
    } else if (textPositionG(position, node.pos.end)) {
        diags.push({
            errorType: DiagnoseErrorType.ApplyParserError,
            severity: DiagnoseSeverity.Fatal,
            pos: {
                start: node.pos.end,
                end: position,
            },
            message: 'Parser position overflow',
            filePath: nodeE.parentFile?.path ?? 'null',
        });
    }

    return [newNodes, diags];
}

type Visitor<T> = (node: RawNodeE) => T[];
const parsers: {
    [key in LexemeType]: Visitor<LexemeTypeToNodeType[key]>;
} = {
    [LexemeType.Code]: node => {
        const lines = node.text.lines;

        const codeLexemes = lines
            .map((v, i) => ({
                offLine: i,
                absLine: node.n.pos.start.line + i,
                match: v.s.match(/\s*`{3,}/),
            }))
            .filter(v => v.match?.length);
        if (codeLexemes.length % 2 !== 0) {
            const lastLexeme = codeLexemes[codeLexemes.length - 1];
            const absLine = lastLexeme.absLine;
            throw new Error(
                `Unable to find closing quotes for block code. ` +
                `Began at ${absLine} at file TODO`,
            );
        }

        const codeNodes: CodeNode[] = [];
        for (let i = 0; i < codeLexemes.length; i += 2) {
            const startLexeme = codeLexemes[i];
            const endLexeme = codeLexemes[i + 1];

            const text = lines
                .slice(startLexeme.offLine + 1, endLexeme.offLine)
                .join('\n');
            codeNodes.push({
                type: NodeType.Code,
                pos: createStartEndPos(
                    startLexeme.absLine,
                    1,
                    endLexeme.absLine,
                    lines[endLexeme.offLine].length,
                ),
                text: text,
                parent: node.n.parent,
                lang: null,
                codeBlockStyle: null,
            });
        }

        return codeNodes;
    },
};
