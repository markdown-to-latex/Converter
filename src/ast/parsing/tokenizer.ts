/**
 * Used for selecting certain tokens
 */
import { RawNodeType, TokensNode } from '../node';

export interface Token {
    text: string;
    pos: number;
    type: TokenType;
}

export interface TokenizeResult {
    text: string;
    tokens: Token[];
    basePos: number;
}

export const enum TokenType {
    Spacer = 'Spacer' /* Spaces, tabs, line breaks */,
    Letter = 'Letter' /* Any language letter and numbers */,
    SeparatedSpecial = 'SeparatedSpecial' /* Brackets, semicolons */,
    JoinableSpecial = 'JoinableSpecial' /* Octothorpe, backticks */,
    Delimiter = 'Delimiter' /* Line break */,
    Other = 'Other' /* Not matched characters */,
}

const JOINABLE_TOKEN_TYPES: TokenType[] = [
    TokenType.Spacer,
    TokenType.Letter,
    TokenType.Delimiter,
];

const SPACER_REGEXP = new RegExp(/[ \t\r]/);
const SEPARATED_SPECIAL_REGEXP = new RegExp(
    /[\[\]{}~!@%^&()=+\\,.<>;:'"|â„–?]/,
);
const JOINABLE_SPECIAL_REGEXP = new RegExp(/[`#$*\-_\/]/);
const NUMBER_REGEXP = new RegExp(/\d/);

function isCharSeparatedSpecial(char: string): boolean {
    return !!char.match(SEPARATED_SPECIAL_REGEXP);
}

function isCharJoinableSpecial(char: string): boolean {
    return !!char.match(JOINABLE_SPECIAL_REGEXP);
}

function isCharSpacer(char: string): boolean {
    return !!char.match(SPACER_REGEXP);
}

function isCharLetter(char: string): boolean {
    return char.toLowerCase() !== char.toUpperCase();
}

function isCharNumber(char: string): boolean {
    return !!char.match(NUMBER_REGEXP);
}

function isCharDelimiter(char: string): boolean {
    return char === '\n';
}

function getCharType(char: string): TokenType {
    if (isCharDelimiter(char)) {
        return TokenType.Delimiter;
    }
    if (isCharSpacer(char)) {
        return TokenType.Spacer;
    }
    if (isCharJoinableSpecial(char)) {
        return TokenType.JoinableSpecial;
    }
    if (isCharSeparatedSpecial(char)) {
        return TokenType.SeparatedSpecial;
    }
    if (isCharLetter(char) || isCharNumber(char)) {
        return TokenType.Letter;
    }

    return TokenType.Other;
}

function createToken(char: string, pos: number): Token {
    return {
        text: char,
        pos: pos,
        type: getCharType(char),
    };
}

export function tokenize(text: string, basePos: number = 0): TokenizeResult {
    const tokens: Token[] = [];
    let currentToken: Token | null = null;

    for (const [i, c] of text.split('').map((v, i) => [i, v] as const)) {
        if (!currentToken) {
            currentToken = createToken(c, i + basePos);
            continue;
        }

        const type = getCharType(c);
        const currentTokenEndChar =
            currentToken.text[currentToken.text.length - 1];
        const joinables = ['$', '`'];
        if (
            type === currentToken.type &&
            JOINABLE_TOKEN_TYPES.indexOf(type) !== -1
        ) {
            currentToken.text += c;
        } else if (
            type === TokenType.JoinableSpecial &&
            (currentTokenEndChar === c ||
                (joinables.indexOf(currentTokenEndChar) !== -1 &&
                    joinables.indexOf(c) !== -1))
        ) {
            currentToken.text += c;
        } else {
            tokens.push(currentToken);
            currentToken = createToken(c, i + basePos);
        }
    }

    if (currentToken) {
        tokens.push(currentToken);
    }

    return {
        text,
        basePos,
        tokens: tokens,
    };
}

export function tokensToNode(tokens: TokenizeResult): TokensNode {
    let lastToken: Token | null = tokens.tokens.length
        ? tokens.tokens[tokens.tokens.length - 1]
        : null;
    return {
        type: RawNodeType.Tokens,
        text: tokens.text,
        tokens: tokens.tokens,
        pos: {
            start: tokens.basePos,
            end: lastToken
                ? lastToken.pos + lastToken.text.length
                : tokens.basePos,
        },
        parent: null,
    };
}
