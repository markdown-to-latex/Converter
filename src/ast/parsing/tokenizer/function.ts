import { RawNodeType, TokensNode } from '../../node';

export interface Token {
    text: string;
    pos: number;
    type: TokenType;
}

export interface TokenizeResult {
    text: string;
    tokens: Token[];
    basePos: number;
}

export const enum TokenType {
    Spacer = 'Spacer' /* Spaces, tabs, line breaks */,
    SeparatedSpecial = 'SeparatedSpecial' /* Brackets, semicolons */,
    JoinableSpecial = 'JoinableSpecial' /* Octothorpe, backticks */,
    Delimiter = 'Delimiter' /* Line break */,
    Letter = 'Letter' /* Not parsed. Any language letter and numbers */,
}

export interface RegExpTokenData {
    tokenType: TokenType;
    charRegExpStr: string;
    joinable: boolean;
}

const REGEXP_CHARS = {
    spacer: ' \\t',
    separatedSpecial: '\\[\\]{}!@%\\^&()+\\\\,.\\;:\'"|â„–?',
    joinableSpecial: '#*\\-_\\/~=<>',
    doubleJoinableSpecial: '$`',
    delimiter: '\\r\\n',
} as const;

type RegExpTypes = keyof typeof REGEXP_CHARS | 'letter';

const REGEXP_TOKEN_DATA_BY_GROUP_NAME: Record<RegExpTypes, RegExpTokenData> = {
    spacer: {
        tokenType: TokenType.Spacer,
        charRegExpStr: `[${REGEXP_CHARS.spacer}]`,
        joinable: true,
    },
    separatedSpecial: {
        tokenType: TokenType.SeparatedSpecial,
        charRegExpStr: `[${REGEXP_CHARS.separatedSpecial}]`,
        joinable: false,
    },
    joinableSpecial: {
        tokenType: TokenType.JoinableSpecial,
        charRegExpStr: `(?<__jSpecialIn>[${REGEXP_CHARS.joinableSpecial}])\\k<__jSpecialIn>*`,
        joinable: false,
    },
    doubleJoinableSpecial: {
        tokenType: TokenType.JoinableSpecial,
        charRegExpStr: `[${REGEXP_CHARS.doubleJoinableSpecial}]`,
        joinable: true,
    },
    delimiter: {
        tokenType: TokenType.Delimiter,
        charRegExpStr: `[${REGEXP_CHARS.delimiter}]`,
        joinable: true,
    },
    letter: {
        tokenType: TokenType.Letter,
        charRegExpStr: `[^${REGEXP_CHARS.spacer}${REGEXP_CHARS.separatedSpecial}${REGEXP_CHARS.joinableSpecial}${REGEXP_CHARS.doubleJoinableSpecial}${REGEXP_CHARS.delimiter}]`,
        joinable: true,
    },
} as const;

const TOKENIZING_REGEXP = new RegExp(
    Object.entries(REGEXP_TOKEN_DATA_BY_GROUP_NAME)
        .map(([key, value]) => {
            const multiplier = value.joinable ? '+' : '';
            return `(?<${key}>${value.charRegExpStr}${multiplier})`;
        })
        .join('|'),
    'g',
);
console.log(TOKENIZING_REGEXP);

export function tokenize(text: string, basePos: number = 0): TokenizeResult {
    const tokens = Array.from(text.matchAll(TOKENIZING_REGEXP)).flatMap(
        result => {
            const keyResult = Object.entries(REGEXP_TOKEN_DATA_BY_GROUP_NAME)
                .map(([k, v]) => [k, v, result.groups?.[k]] as const)
                .find<[string, RegExpTokenData, string]>(
                    (
                        prev: readonly [
                            string,
                            RegExpTokenData,
                            string | undefined,
                        ],
                    ): prev is [string, RegExpTokenData, string] => !!prev[2],
                );
            if (!keyResult) {
                return [];
            }

            const [key, data, regexpResult] = keyResult;
            return [
                {
                    type: data.tokenType,
                    pos: result.index,
                    text: regexpResult,
                },
            ] as Token[];
        },
    );

    return {
        text,
        basePos,
        tokens: tokens,
    };
}

export function tokensToNode(tokens: TokenizeResult): TokensNode {
    let lastToken: Token | null = tokens.tokens.length
        ? tokens.tokens[tokens.tokens.length - 1]
        : null;
    return {
        type: RawNodeType.Tokens,
        text: tokens.text,
        tokens: tokens.tokens,
        pos: {
            start: tokens.basePos,
            end: lastToken
                ? lastToken.pos + lastToken.text.length
                : tokens.basePos,
        },
        parent: null,
    };
}
