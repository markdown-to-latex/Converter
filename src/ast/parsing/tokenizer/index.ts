/**
 * Used for selecting certain tokens
 */

export * from './function';
export * from './utils';
