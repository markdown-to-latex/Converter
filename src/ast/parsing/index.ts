import {
    createStartEndTextPos,
    FileNode,
    NodeE,
    NodeType,
    RawNode,
    RawNodeType,
    StartEndTextPosition,
} from '../node';
import path from 'path';
import { StringE } from '../../extension/string';
import { tokenize, tokensToNode } from './tokenizer';

export function fullContentPos(
    content: string | StringE,
): StartEndTextPosition {
    const contentE = StringE.from(content);

    const lines = contentE.splitE(/\r?\n/);
    return createStartEndTextPos(
        1,
        1,
        lines.length,
        lines[lines.length - 1].length + 1,
    );
}

function parseFile(content: string, filePath: string) {
    const contentNode: RawNode = {
        type: RawNodeType.Raw,
        parent: null,
        text: content,
        pos: {
            start: 0,
            end: content.length,
        },
    };

    const fileNode: FileNode = {
        type: NodeType.File,
        parent: null,
        children: [contentNode],
        pos: {
            start: 0,
            end: content.length,
        },
        raw: content,
        path: path.resolve(filePath),
    };

    contentNode.parent = fileNode;

    let nodeE = NodeE.from(fileNode);
}
