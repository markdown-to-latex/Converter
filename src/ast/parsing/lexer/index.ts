import * as node from "../../node";
import {
    getNodeParentFile,
    Node,
    NodeChildren,
    NodeText,
    NodeType,
    positionToTextPosition,
    RawNode,
    RawNodeType,
    TextNode,
    TokensNode
} from "../../node";
import {
    DiagnoseErrorType,
    DiagnoseInfo,
    DiagnoseList,
    diagnoseListHasSeverity,
    DiagnoseSeverity
} from "../../../diagnose";
import { Token, tokenize, tokensToNode, TokenType } from "../tokenizer";
import { TokenByTypeParserResult, TokenParser, TokenPredicate } from "./struct";
import { parseCode } from "./node/code";
import {
    isParagraphBreak,
    parseParagraphBreak,
    parseSoftBreak
} from "./node/paragraph";
import { parseCodeSpan } from "./node/codeSpan";
import { parseLink } from "./node/link";
import { parseMacro } from "./node/macros";
import { parseTable } from "./node/table";
import { parseList } from "./node/list";
import { parseHeading } from "./node/heading";
import { parseBlockquote } from "./node/blockquote";
import { parseImage } from "./node/image";
import { parseHr } from "./node/hr";
import { parseEm } from "./node/em";
import { parseStrongWithOptionalEm } from "./node/strong";

export const enum LexemeType {
    // Space,
    // Paragraph,
    // Heading,
    // Text,
    Code,
}

const parserPriorities: LexemeType[] = [
    LexemeType.Code
    // LexemeType.Heading,
    // LexemeType.Paragraph,
    // LexemeType.Text,
    // LexemeType.Space,
];

interface LexemeTypeToNodeType {
    [LexemeType.Code]: node.CodeNode;
    // [LexemeType.Heading]: node.HeadingNode;
    // [LexemeType.Paragraph]: node.ParagraphNode;
    // [LexemeType.Text]: node.TextNode;
    // [LexemeType.Space]: node.SpaceNode;
}

class FatalError extends Error {
}

interface ApplyVisitorsResult {
    nodes: Node[];
    diagnostic: DiagnoseList;
}

export function applyVisitors(nodes: Readonly<Node[]>): ApplyVisitorsResult {
    const visitorsResult: ApplyVisitorsResult = {
        nodes: [...nodes],
        diagnostic: []
    };

    try {
        for (const type of parserPriorities) {
            visitorsResult.nodes = visitorsResult.nodes.flatMap(node => {
                if (node.type === RawNodeType.Raw) {
                    const parent = node.parent;
                    node = tokensToNode(tokenize((node as RawNode).text, 0));
                    node.parent = parent;
                }

                if (node.type !== RawNodeType.Tokens) {
                    return [node];
                }

                const result = parseTokensNode(node as TokensNode);
                visitorsResult.diagnostic.push(...result.diagnostic);

                if (
                    diagnoseListHasSeverity(
                        result.diagnostic,
                        DiagnoseSeverity.Fatal
                    )
                ) {
                    throw new FatalError();
                }

                return result.nodes;
            });
        }
    } catch (e) {
        if (e instanceof FatalError) {
            return visitorsResult;
        } else {
            throw e;
        }
    }

    return visitorsResult;
}

interface FindTokenResult {
    token: Token;
    index: number;
}

export function findTokenOrNull(
    tokens: TokensNode,
    fromIndex: number,
    predicate: TokenPredicate
): FindTokenResult | null {
    for (let i = fromIndex; i < tokens.tokens.length; ++i) {
        const token = tokens.tokens[i];
        if (predicate(token, i, tokens)) {
            return {
                token,
                index: i
            };
        }
    }
    return null;
}

export function findTokenOrNullBackward(
    tokens: TokensNode,
    fromIndex: number,
    predicate: TokenPredicate
): FindTokenResult | null {
    for (let i = fromIndex; i >= 0; --i) {
        const token = tokens.tokens[i];
        if (predicate(token, i, tokens)) {
            return {
                token,
                index: i
            };
        }
    }
    return null;
}

const brackets: Record<string, string> = {
    "[": "]", // Label
    "{": "}", // Reserved
    "(": ")", // Argument
    "`": "`", // Code Span
    "$`": "`$" // Formula Span
};

const bracketsBackward: Record<string, string> = Object.fromEntries(
    Object.entries(brackets).map(([k, v]) => [v, k])
);

const exclusiveBrackets = ["`", "$`"];

export function findTokenClosingBracket(
    tokens: TokensNode,
    index: number,
    greedy: boolean = false /* Capture self-closing bracket only */
): FindTokenResult | null {
    const openToken = tokens.tokens[index];
    const closeBracket = brackets[openToken.text] ?? openToken.text;

    const bracketsCounter: Record<string, number> = Object.fromEntries(
        Object.keys(brackets).map(k => [k, 0])
    );

    for (let i = index + 1; i < tokens.tokens.length; ++i) {
        const token = tokens.tokens[i];
        if (isParagraphBreak(token, i, tokens)) {
            return null; // break breaks the brackets
        }

        if (
            token.type !== TokenType.JoinableSpecial &&
            token.type !== TokenType.SeparatedSpecial
        ) {
            continue;
        }

        if (greedy) {
            if (token.text === closeBracket) {
                return {
                    token: token,
                    index: i
                };
            }
        } else {
            if (
                Object.values(bracketsCounter).reduce(
                    (prev, current) => prev + current,
                    0
                ) === 0
            ) {
                if (token.text === closeBracket) {
                    return {
                        token: token,
                        index: i
                    };
                }
            }

            const exclusiveBracket =
                exclusiveBrackets
                    .map(v => [v, bracketsCounter[v]] as const)
                    .find(v => v[1]) ?? null;
            if (exclusiveBracket) {
                const exclusiveOpenBracket = exclusiveBracket[0];
                const exclusiveCloseBracket = brackets[exclusiveOpenBracket];
                if (token.text == exclusiveCloseBracket) {
                    bracketsCounter[exclusiveOpenBracket] -= 1;
                }
            } else {
                // Count all inner brackets
                if (token.text in brackets) {
                    bracketsCounter[token.text] += 1;
                } else if (token.text in bracketsBackward) {
                    bracketsCounter[bracketsBackward[token.text]] -= 1;
                }
            }
        }
    }

    return null;
}

export function isOpenLabelBracket(tokens: TokensNode, index: number): boolean {
    const token = tokens.tokens[index];

    if (
        token.type !== TokenType.JoinableSpecial &&
        token.type !== TokenType.SeparatedSpecial
    ) {
        return false;
    }

    return token.text == "[";
}

export function isOpenArgumentBracket(
    tokens: TokensNode,
    index: number
): boolean {
    const token = tokens.tokens[index];

    if (
        token.type !== TokenType.JoinableSpecial &&
        token.type !== TokenType.SeparatedSpecial
    ) {
        return false;
    }

    return token.text == "(";
}

interface TrimWithPositionsResult {
    result: string;
    leftTrim: number;
    rightTrim: number;
}

export function trimSingleWithPositions(str: string): TrimWithPositionsResult {
    const originalLen = str.length;
    const leftLen = originalLen - str.trimStart().length;
    const rightLen = originalLen - str.trimEnd().length;

    if (originalLen <= leftLen + rightLen) {
        return {
            result: " ",
            leftTrim: originalLen - 1,
            rightTrim: 0
        };
    }

    return {
        result: str.slice(
            leftLen > 0 ? leftLen - 1 : 0,
            rightLen > 0 ? originalLen - rightLen + 1 : originalLen
        ),
        leftTrim: leftLen > 0 ? leftLen - 1 : 0,
        rightTrim: rightLen > 0 ? rightLen - 1 : 0
    };
}

interface ParseTokensNodeResult {
    nodes: Node[];
    diagnostic: DiagnoseList;
}

export function parseTokensNode(tokens: TokensNode): ParseTokensNodeResult {
    const parsingResult: ParseTokensNodeResult = {
        nodes: [],
        diagnostic: []
    };

    for (let i = 0; i < tokens.tokens.length /*manual*/;) {
        const token = tokens.tokens[i];

        const result = parseTokensNodeByType(tokens, i);
        if (result) {
            i = result.index;
            parsingResult.diagnostic.push(...result.diagnostic);
            parsingResult.nodes.push(...result.nodes);

            if (
                diagnoseListHasSeverity(
                    result.diagnostic,
                    DiagnoseSeverity.Fatal
                )
            ) {
                return parsingResult;
            }

            continue;
        }

        parsingResult.nodes.push({
            type: NodeType.Text,
            parent: tokens.parent,
            text: token.text,
            pos: {
                start: token.pos,
                end: token.pos + token.text.length
            },
            children: []
        } as TextNode);
        ++i;
    }

    nodeJoiner(parsingResult.nodes);
    // parsingResult.nodes = parsingResult.nodes.map(n => {
    //     if (n.type === NodeType.Text) {
    //         const trimmed = trimSingleWithPositions((n as TextNode).text);
    //         return {
    //             type: NodeType.Text,
    //             text: trimmed.result,
    //             parent: n.parent,
    //             pos: {
    //                 start: n.pos.start + trimmed.leftTrim,
    //                 end: n.pos.end - trimmed.rightTrim,
    //             },
    //             children: (n as TextNode).children,
    //         } as TextNode;
    //     }
    //     return n;
    // });
    parsingResult.nodes = parsingResult.nodes.filter(
        n => n.type !== RawNodeType.ParagraphBreak
    );
    return parsingResult;
}

function nodeJoiner(nodes: Node[]): void {
    const breaks: (NodeType | RawNodeType)[] = [
        RawNodeType.SoftBreak,
        RawNodeType.ParagraphBreak
    ];
    if (breaks.indexOf(nodes[0]?.type) !== -1) {
        nodes.splice(0, 1);
    }

    // TODO: Trim nodes
    while (nodes.length) {
        const textNode = nodes[0] as TextNode;

        if (
            textNode.type !== NodeType.Text ||
            textNode.text.trim().length !== 0
        ) {
            break;
        }

        nodes.splice(0, 1);
    }
    while (nodes.length) {
        const textNode = nodes[nodes.length - 1] as TextNode;

        if (
            textNode.type !== NodeType.Text ||
            textNode.text.trim().length !== 0
        ) {
            break;
        }

        nodes.splice(nodes.length - 1, 1);
    }

    let index = 1;
    while (index < nodes.length) {
        const currentNode = nodes[index];
        const prevNode = nodes[index - 1];

        if (currentNode.type === RawNodeType.SoftBreak) {
            if (
                prevNode.type === NodeType.Text &&
                nodes[index + 1]?.type === NodeType.Text
            ) {
                const prevNodeText = prevNode as Node & NodeText;

                prevNodeText.pos.end = currentNode.pos.end;
                prevNodeText.text += " ";
            }

            nodes.splice(index, 1);
            continue;
        }

        if (
            currentNode.type === prevNode.type &&
            currentNode.type === NodeType.Text
        ) {
            // Contains 'text' field => is a NodeText
            const currentNodeText = currentNode as Node & NodeText;
            const prevNodeText = prevNode as Node & NodeText;

            prevNodeText.pos.end = currentNodeText.pos.end;
            prevNodeText.text += currentNodeText.text;

            nodes.splice(index, 1);
            continue;
        }

        if (
            currentNode.type === prevNode.type &&
            currentNode.type === NodeType.Blockquote
        ) {
            // Contains 'text' field => is a NodeText
            const currentNodeChildren = currentNode as Node & NodeChildren;
            const prevNodeChildren = prevNode as Node & NodeChildren;

            prevNodeChildren.pos.end = currentNodeChildren.pos.end;
            prevNodeChildren.children.push(...currentNodeChildren.children);
            currentNodeChildren.children.forEach(
                n => (n.parent = prevNodeChildren)
            );

            nodes.splice(index, 1);
            continue;
        }

        ++index;
    }
}

function parseTokensNodeByType(
    tokens: TokensNode,
    index: number
): TokenByTypeParserResult | null {
    const token = tokens.tokens[index];

    for (const parser of parsersByType[token.type]) {
        const result = parser(tokens, index);
        if (result) {
            return result;
        }
    }

    return null;
}

const parsersByType: Record<TokenType, TokenParser[]> = {
    [TokenType.JoinableSpecial]: [
        parseCodeSpan,
        parseStrongWithOptionalEm,
        parseEm,
        parseHeading,
        parseHr,
        parseCode
    ],
    [TokenType.SeparatedSpecial]: [
        parseLink,
        parseMacro,
        parseImage,
        parseTable,
        parseBlockquote
    ],
    [TokenType.Delimiter]: [parseSoftBreak, parseParagraphBreak],
    [TokenType.Spacer]: [parseList],
    [TokenType.Letter]: [parseList],
    [TokenType.Other]: []
};

export function tokenToDiagnose(
    tokens: TokensNode,
    index: number,
    message: string,
    severity: DiagnoseSeverity = DiagnoseSeverity.Fatal
): DiagnoseInfo {
    const token = tokens.tokens[index];
    const nodeParentFile = getNodeParentFile(tokens);
    return {
        message: message,
        severity: severity,
        filePath: nodeParentFile?.path ?? "null",
        errorType: DiagnoseErrorType.ApplyParserError,
        pos: {
            start: {
                absolute: token.pos,
                ...positionToTextPosition(nodeParentFile?.raw ?? "", token.pos)
            },
            end: {
                absolute: token.pos + token.text.length,
                ...positionToTextPosition(
                    nodeParentFile?.raw ?? "",
                    token.pos + token.text.length
                )
            }
        }
    };
}

export function unexpectedEof(
    tokens: TokensNode,
    index: number,
    message: string
): TokenByTypeParserResult {
    return {
        nodes: [],
        index: tokens.tokens.length,
        diagnostic: [
            tokenToDiagnose(tokens, index, message, DiagnoseSeverity.Error)
        ]
    };
}

export function sliceTokenText(
    tokens: TokensNode,
    fromIndex: number,
    toIndex: number
): string {
    const tokenStart = tokens.tokens[fromIndex];
    const tokenEnd: Token | null = tokens.tokens[toIndex];

    return tokens.text.slice(
        tokenStart.pos - tokens.pos.start,
        tokenEnd ? tokenEnd.pos - tokens.pos.end : tokens.pos.end
    );
}
