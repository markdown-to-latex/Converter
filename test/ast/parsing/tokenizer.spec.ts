import { tokenize } from '../../../src/ast/parsing/tokenizer';

describe('tokenize', function () {
    test('simple text', () => {
        const TEXT = 'Simple text with words and423 numbers;\n';
        const tokens = tokenize(TEXT);

        expect(tokens).toMatchSnapshot();
    });
    test('Joined Tokens', () => {
        const TEXT = '![message,](simple-text2-#)';
        const tokens = tokenize(TEXT);

        expect(tokens).toMatchSnapshot();
    });
    test('Multiline', () => {
        const TEXT = 'Line 1\nLine 2 [] \n  Line{$$ 3  \nLine4.';
        const tokens = tokenize(TEXT);

        expect(tokens).toMatchSnapshot();
    });
});
