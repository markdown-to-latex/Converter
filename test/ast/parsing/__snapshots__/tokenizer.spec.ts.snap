// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize Joined Tokens 1`] = `
Object {
  "basePos": 0,
  "text": "![message,](simple-text2-#)",
  "tokens": Array [
    Object {
      "pos": 0,
      "text": "!",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 1,
      "text": "[",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 2,
      "text": "message",
      "type": "Letter",
    },
    Object {
      "pos": 9,
      "text": ",",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 10,
      "text": "]",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 11,
      "text": "(",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 12,
      "text": "simple",
      "type": "Letter",
    },
    Object {
      "pos": 18,
      "text": "-",
      "type": "JoinableSpecial",
    },
    Object {
      "pos": 19,
      "text": "text2",
      "type": "Letter",
    },
    Object {
      "pos": 24,
      "text": "-#",
      "type": "JoinableSpecial",
    },
    Object {
      "pos": 26,
      "text": ")",
      "type": "SeparatedSpecial",
    },
  ],
}
`;

exports[`tokenize Multiline 1`] = `
Object {
  "basePos": 0,
  "text": "Line 1
Line 2 [] 
  Line{$$ 3  
Line4.",
  "tokens": Array [
    Object {
      "pos": 0,
      "text": "Line",
      "type": "Letter",
    },
    Object {
      "pos": 4,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 5,
      "text": "1",
      "type": "Letter",
    },
    Object {
      "pos": 6,
      "text": "
",
      "type": "Delimiter",
    },
    Object {
      "pos": 7,
      "text": "Line",
      "type": "Letter",
    },
    Object {
      "pos": 11,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 12,
      "text": "2",
      "type": "Letter",
    },
    Object {
      "pos": 13,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 14,
      "text": "[",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 15,
      "text": "]",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 16,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 17,
      "text": "
",
      "type": "Delimiter",
    },
    Object {
      "pos": 18,
      "text": "  ",
      "type": "Spacer",
    },
    Object {
      "pos": 20,
      "text": "Line",
      "type": "Letter",
    },
    Object {
      "pos": 24,
      "text": "{",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 25,
      "text": "$$",
      "type": "JoinableSpecial",
    },
    Object {
      "pos": 27,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 28,
      "text": "3",
      "type": "Letter",
    },
    Object {
      "pos": 29,
      "text": "  ",
      "type": "Spacer",
    },
    Object {
      "pos": 31,
      "text": "
",
      "type": "Delimiter",
    },
    Object {
      "pos": 32,
      "text": "Line4",
      "type": "Letter",
    },
    Object {
      "pos": 37,
      "text": ".",
      "type": "SeparatedSpecial",
    },
  ],
}
`;

exports[`tokenize simple text 1`] = `
Object {
  "basePos": 0,
  "text": "Simple text with words and423 numbers;
",
  "tokens": Array [
    Object {
      "pos": 0,
      "text": "Simple",
      "type": "Letter",
    },
    Object {
      "pos": 6,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 7,
      "text": "text",
      "type": "Letter",
    },
    Object {
      "pos": 11,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 12,
      "text": "with",
      "type": "Letter",
    },
    Object {
      "pos": 16,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 17,
      "text": "words",
      "type": "Letter",
    },
    Object {
      "pos": 22,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 23,
      "text": "and423",
      "type": "Letter",
    },
    Object {
      "pos": 29,
      "text": " ",
      "type": "Spacer",
    },
    Object {
      "pos": 30,
      "text": "numbers",
      "type": "Letter",
    },
    Object {
      "pos": 37,
      "text": ";",
      "type": "SeparatedSpecial",
    },
    Object {
      "pos": 38,
      "text": "
",
      "type": "Delimiter",
    },
  ],
}
`;
